test/e2e/helpers.go

```diff
@@ -911,3 +913,77 @@ func WaitForDaemonset(ctx context.Context, input clusterctl.ApplyClusterTemplate
 		return false
 	}, input.WaitForControlPlaneIntervals...).Should(Equal(true))
 }
+
+// podListHasNumPods fulfills the Cluster API PodListCondition type spec
+// given a list of pods, we validate for an exact number of those pods in a Running state
+func podListHasNumPods(numPods int) func(pl *corev1.PodList) error {
+	return func(pl *corev1.PodList) error {
+		var runningPods int
+		for _, p := range pl.Items {
+			if p.Status.Phase == corev1.PodRunning {
+				runningPods++
+			}
+		}
+		if runningPods != numPods {
+			return errors.Errorf("expected %d Running pods, got %d", numPods, runningPods)
+		}
+		return nil
+	}
+}
+
+// podListHasAtLeastNumPods fulfills the cluster-api PodListCondition type spec
+// given a list of pods, we validate for at least 1 number of those pods in a Running state
+func podListHasAtLeastNumPods(numPods int) func(pl *corev1.PodList) error {
+	return func(pl *corev1.PodList) error {
+		var runningPods int
+		for _, p := range pl.Items {
+			if p.Status.Phase == corev1.PodRunning {
+				runningPods++
+			}
+		}
+		if runningPods >= numPods {
+			return errors.Errorf("expected %d Running pods, got %d", numPods, runningPods)
+		}
+		return nil
+	}
+}
+
+func defaultConfigCluster(clusterName, namespace string) clusterctl.ConfigClusterInput {
+	return clusterctl.ConfigClusterInput{
+		LogFolder:                filepath.Join(artifactFolder, "clusters", bootstrapClusterProxy.GetName()),
+		ClusterctlConfigPath:     clusterctlConfigPath,
+		KubeconfigPath:            bootstrapClusterProxy.GetKubeconfigPath(),
+		InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+		Flavor:                   clusterctl.DefaultFlavor,
+		Namespace:                namespace,
+		ClusterName:              clusterName,
+		KubernetesVersion:        e2eConfig.GetVariable(capi_e2e.KubernetesVersion),
+		ControlPlaneMachineCount: pointer.Int64Ptr(3),
+		WorkerMachineCount:       pointer.Int64Ptr(0),
+	}
+}
+
+func createCluster(ctx context.Context, configCluster clusterctl.ConfigClusterInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) (*clusterv1.Cluster, []*clusterv1.MachineDeployment, *controlplanev1.KubeadmControlPlane) {
```

Comment:
Why not reuse the below function with a nil controlplane waiters parameter instead of creating a separate function?

Reply:
Makes sense.